### PR TITLE
Intersection over Union Metric/Loss

### LINKED ISSUE #370: Add Intersection over Union loss
## 🚀 Feature

Since #251 includes GIoU loss, I think having IoU loss only makes sense.

### Pitch

I can implement this pretty quickly since I've been working on GIoU #347.

### PR DESCRIPTION
**Pull request by briankosw**
## What does this PR do?

Implements IoU metric and IoU loss

Fixes #370. Related to #347 and #251.

## Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning-bolts/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? [not needed for typos/docs]
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning-bolts/blob/master/CHANGELOG.md)?

## PR review
 - [x] Is this pull request ready for review? (if not, please submit in draft mode)

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by oke-aditya** [comment]
But IoU and gIoU are available in torchvision. Thoughts ?

File: pl_bolts/metrics/object_detection.py [hunk 1]
```diff
@@ -1,6 +1,38 @@
...
+    y_max = torch.min(preds[:, None, 3], target[:, 3])
+    intersection = (x_max - x_min).clamp(min=0) * (y_max - y_min).clamp(min=0)
+    pred_area = (preds[:, 2] - preds[:, 0]) * (preds[:, 3] - preds[:, 1])
+    target_area = (target[:, 2] - target[:, 0]) * (target[:, 3] - target[:, 1])
+    union = pred_area[:, None] + target_area - intersection
+    iou = torch.true_divide(intersection, union)
+    return iou
+
+
 def giou(preds: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
```
Comment (oke-aditya) [inline pl_bolts/metrics/object_detection.py:36]: Even giou is implemented in torchvision

**Comment by briankosw** [comment]
> But IoU and gIoU are available in torchvision. Thoughts ?
> I agree that gIoU loss IoU loss is not available in torchvision (hopefully soon) but for ops, there is tested support.
> 
> cc @akihironitta maybe we are reimplementing these in bolts to avoid dependency ?

The conversation regarding `pl_bolts` having its own implementation of `giou` and `giou_loss` was held in #347. The conclusion was that it's helpful to have a separate version, hence this PR.

**Comment by oke-aditya** [comment]
Hmm, not an issue then but we would need to test to match torchvision's implementation exactly.
E.g. IoU and gIoU in torchvision return a `Tensor[N, 4]` while here we do a little bit different.
Which might be slightly confusing for users.

It can be nice to have drop-in replacement for consistency.  Thoughts ?

cc @akihironitta @Borda

**Comment by briankosw** [comment]
> Hmm, not an issue then but we would need to test to match torchvision's implementation exactly.
> E.g. IoU and gIoU in torchvision return a `Tensor[N, 4]` while here we do a little bit different.
> Which might be slightly confusing for users.

So IoU and GIoU in `torchvision` actually doesn't return an Nx4 tensor, it returns an NxM tensor, where N is the dimension of the first set of boxes and M is the dimension of the second set of boxes. In fact my implementation does the same, and you bring up an excellent point that perhaps the docstring should correctly reflect this fact to the users. In addition, I can further demonstrate this by creating tests that reflect this fact. Does that sound ok?

**Comment by oke-aditya** [comment]
Oh yes, it returns Tensor[N, M] Yes, we can even test on same values as we did in torchvision.
Tests in torchvision are [here](https://github.com/pytorch/vision/blob/master/test/test_ops.py#L838)

File: pl_bolts/metrics/object_detection.py [hunk 2]
```diff
@@ -1,6 +1,38 @@
...
+
+    Returns:
+        IoU value
+    """
+    x_min = torch.max(preds[:, None, 0], target[:, 0])
+    y_min = torch.max(preds[:, None, 1], target[:, 1])
+    x_max = torch.min(preds[:, None, 2], target[:, 2])
+    y_max = torch.min(preds[:, None, 3], target[:, 3])
+    intersection = (x_max - x_min).clamp(min=0) * (y_max - y_min).clamp(min=0)
+    pred_area = (preds[:, 2] - preds[:, 0]) * (preds[:, 3] - preds[:, 1])
```
Comment (oke-aditya) [inline pl_bolts/metrics/object_detection.py:29]: thoughts to have a helper `box_area` function as in torchvision ?

**Comment by briankosw** [comment]
Great suggestion @oke-aditya! How should I go about changing the description for gIoU? Do I create a separate PR?

File: pl_bolts/metrics/object_detection.py (hunk 2 above)
Comment (briankosw) [reply pl_bolts/metrics/object_detection.py:29]: I didn't think a separate `box_area` method was warranted since its use case is quite narrow. Should I go ahead and implement it?

**Comment by oke-aditya** [comment]
I'm Not sure about that. cc @akihironitta :smile:

File: pl_bolts/metrics/object_detection.py (hunk 2 above)
Comment (oke-aditya) [reply pl_bolts/metrics/object_detection.py:30]: Maybe lets leave it for now. Create when we really need it?

**Comment by akihironitta** [comment]
> But IoU and gIoU are available in torchvision. Thoughts?
> I agree that gIoU loss IoU loss is not available in torchvision (hopefully soon) but for ops, there is tested support.
> 
> cc @akihironitta maybe we are reimplementing these in bolts to avoid dependency?

Since Bolts already (heavily?) depends on `torchvision`, I think we can import `torchvision`'s (g)iou and use them here. It's also nice that they should be tested well by the torchvision team. In #347, @briankosw asked if we should have own implementation, but it seems (and I'm sorry) that we missed the question. So, for now, let's hear what others think of this. cc: @PyTorchLightning/bolts-contributors @Borda 

> > Is there any difference between this and [this one from torchvision](https://pytorch.org/docs/stable/torchvision/ops.html#torchvision.ops.generalized_box_iou) except that we extract giou from 1 in this?
> 
> Ah I didn't realize torchvision has its own implementation. You're correct in that we subtract giou from 1. Should I change it to use torchvision's giou?

---

> It would nice to demonstrate with docstring as well. Maybe just say
 
@briankosw Let's improve the description of the loss in the docs. Could you work on this in another PR?

**Comment by briankosw** [comment]
I'll pause work on this PR, and I'll create another PR that improves the description of GIoU loss then.

**Comment by briankosw** [comment]
I'll pause work on this PR then. I already updated the docs for IoU metric as @oke-aditya suggested, so what description should I improve? @akihironitta

File: pl_bolts/metrics/object_detection.py [hunk 3]
```diff
@@ -1,6 +1,39 @@
 import torch
 
 
+def iou(preds: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
+    """
+    Calculates the intersection over union.
+
+    Args:
+        preds: batch of prediction bounding boxes with representation ``[x_min, y_min, x_max, y_max]``
+        target: batch of target bounding boxes with representation ``[x_min, y_min, x_max, y_max]``
```
Comment (akihironitta) [inline pl_bolts/metrics/object_detection.py:10]: May we have tensor sizes here as well?

**Review by akihironitta — CHANGES_REQUESTED** [review]
@briankosw I left a suggestion. Would you mind considering it?

**Comment by akihironitta** [comment]
> I'll pause work on this PR then.

Right, let's wait for what other members have to say about https://github.com/PyTorchLightning/pytorch-lightning-bolts/pull/469#issuecomment-748864389.

**Comment by oke-aditya** [comment]
Also a small point, If we had to add say CIoU and DIoU (yeah IoUs are so many)
We could potentially then have a combined function.

```
def iou(type="ciou", boxes1: Tensor, boxe2: Tensor):

```

File: pl_bolts/metrics/object_detection.py (hunk 3 above)
Comment (briankosw) [reply pl_bolts/metrics/object_detection.py]: Done!

**Comment by Borda** [comment]
@akihironitta mind re-review? 🐰

**Review by akihironitta — APPROVED** [review]
LGTM!

**Review by Borda — APPROVED** [review]
